solutions/joi-19-cake-circle.mdx

```diff
@@ -0,0 +1,180 @@
+---
+id: joi-19-cake-circle
+source: JOI 2019
+title: Cake Circle
+author: Dustin Miao
+---
+
+# Explanation
+
+First, given some set of cakes $S$, it is always optimal to arrange them in sorted order by color. To prove this, consider inductively adding the cakes in ascending order. It can be shown that placing the cake between the greatest one added so far and the smallest one is always optimal. Thus, the total color is $2(\max(S) - \min(S))$. 
+
+Sort the cakes by color. For any set of cakes, the smallest is the leftmost one and the largest is the rightmost one. The task now becomes choosing some range $[l, r]$ that maximizes the sum of the $m$ largest beauty values, minus $2 \cdot (C_r - C_l)$.
+
+Let $\texttt{opt}(l)$ be the optimal right endpoint for some left endpoint. We can observe that $\texttt{opt}$ is a non-decreasing function (the proof of this has been left as an exercise to the reader). This motivates a divide and conquer solution. 
```

Comment:
how did i not notice this
WHY IS THIS LEFT AS AN EXERCISE TO THE READER

Reply:
because I'm too lazy to write a proof
also it's obvious that it's true, but proving it is a pain

Reply:
jesus christ you're too lazy to do anything i ask you